components/places/src/storage/bookmarks/public.rs

```diff
@@ -0,0 +1,582 @@
+}
+
+fn get_child_guids(db: &PlacesDb, parent: RowId) -> Result<Vec<SyncGuid>> {
+    Ok(db.query_rows_into(
+        "SELECT guid FROM moz_bookmarks
+         WHERE parent = :parent
+         ORDER BY position ASC",
+        &[(":parent", &parent)],
+        |row| row.get_checked(0),
+    )?)
+}
+
+fn fetch_bookmark_child_info(
+    db: &PlacesDb,
+    parent: &RawBookmark,
+    get_direct_children: bool,
+    scope: &crate::db::InterruptScope,
+) -> Result<(Option<Vec<SyncGuid>>, Option<Vec<PublicNode>>)> {
+    if parent.bookmark_type != BookmarkType::Folder {
+        return Ok((None, None));
+    }
+    // If we already know that we have no children, don't
+    // bother querying for them.
+    if parent.child_count == 0 {
+        return Ok(if get_direct_children {
+            (None, Some(vec![]))
+        } else {
+            (Some(vec![]), None)
+        });
+    }
+    if !get_direct_children {
+        // Just get the guids.
+        return Ok((Some(get_child_guids(db, parent.row_id)?), None));
+    }
+    // Fetch children. the future this should probably be done by allowing a
+    // depth parameter to be passed into fetch_tree.
+    let child_nodes = get_raw_bookmarks_with_parent(db, parent.row_id)?
+        .into_iter()
+        .map(|kid| {
+            let child_guids = if kid.bookmark_type == BookmarkType::Folder {
+                if kid.child_count == 0 {
+                    Some(vec![])
+                } else {
+                    Some(get_child_guids(db, kid.row_id)?)
+                }
+            } else {
+                None
+            };
+            scope.err_if_interrupted()?;
+            Ok(PublicNode::from(kid).with_children(child_guids, None))
+        })
+        .collect::<Result<Vec<_>>>()?;
+    Ok((None, Some(child_nodes)))
+}
+
+// Implementation of fetch_bookmark
+fn fetch_bookmark_in_tx(
+    db: &PlacesDb,
+    item_guid: &SyncGuid,
+    get_direct_children: bool,
+    scope: &crate::db::InterruptScope,
+) -> Result<Option<PublicNode>> {
+    // get_raw_bookmark doesn't work for the bookmark root, so we just return None explicitly
+    // (rather than erroring). This isn't ideal, but there's no point to fetching the "true"
+    // bookmark root without fetching its children too, so whatever.
+    let rb = if let Some(raw) = get_raw_bookmark(db, item_guid)? {
+        raw
+    } else {
+        return Ok(None);
+    };
+
+    scope.err_if_interrupted()?;
+    // If we're a folder that has children, fetch child guids or children.
+    let (child_guids, child_nodes) =
+        fetch_bookmark_child_info(db, &rb, get_direct_children, scope)?;
+
+    Ok(Some(
+        PublicNode::from(rb).with_children(child_guids, child_nodes),
+    ))
+}
+
+pub fn update_bookmark_from_message(db: &PlacesDb, msg: ProtoBookmark) -> Result<()> {
+    let info = conversions::BookmarkUpdateInfo::from(msg);
+
+    let tx = db.unchecked_transaction()?;
+    let node_type: BookmarkType = db.query_row_and_then_named(
```

Comment:
It would be nicer to call `get_raw_bookmark` here, then pass it as an argument to `update_bookmark_in_tx`. As it is, we're doing a read for the type, then turning around and reading the same bookmark again in `update_bookmark_in_tx`.

Reply:
I'll look but this will probably be followup material.